src/main/java/frc/robot/subsystems/intake/BallIntakeIO.kt

```diff
@@ -0,0 +1,88 @@
+package frc.robot.subsystems.intake
+
+import com.revrobotics.CANSparkMax
+import com.revrobotics.CANSparkMaxLowLevel
+import com.revrobotics.SparkMaxAbsoluteEncoder
+import edu.wpi.first.math.geometry.Rotation2d
+import edu.wpi.first.math.util.Units
+import org.littletonrobotics.junction.LogTable
+import org.littletonrobotics.junction.inputs.LoggableInputs
+
+interface IntakeIO  {
+
+    class Inputs : LoggableInputs {
+        var position = Rotation2d()
+        var armVelocity = Rotation2d()
+
+            override fun toLog(table: LogTable?) {
+                table?.put("Intake Arm Position", position.radians)
+                table?.put("Intake Arm Speed", armVelocity.radians)
+            }
+
+
+            override fun fromLog(table: LogTable?) {
+                table?.getDouble("Intake Arm Position", position.radians)?.let { position = Rotation2d(it) }
+                table?.getDouble("Intake Arm Speed", armVelocity.radians)?.let { armVelocity = Rotation2d(it) }
+            }
+    }
+
+    fun updateInputs(inputs: Inputs)
+
+
+    fun setArmSpeed(speed: Double)
+
+
+    fun setArmVoltage(outputVolts: Double) {}
+
+    fun setRollerSpeed(speed: Double)
+
+
+    fun setRollerVoltage(outputVolts: Double) {}
+}
+
+
+class IntakeIOReal(ArmMotorID: Int, RollersMotorID: Int) : IntakeIO {
+
+
+
+    private val armMotor = CANSparkMax(ArmMotorID, CANSparkMaxLowLevel.MotorType.kBrushless)
+    private val armEncoder = armMotor.getAbsoluteEncoder(SparkMaxAbsoluteEncoder.Type.kDutyCycle)
+    private val rollerMotor = CANSparkMax(RollersMotorID, CANSparkMaxLowLevel.MotorType.kBrushless)
+
+    init {
+        armEncoder.positionConversionFactor = Units.rotationsToRadians(1.0) *  GEAR_RATIO
+        armEncoder.velocityConversionFactor = Units.rotationsToRadians(1.0) * GEAR_RATIO / 60
+        armMotor.encoder.velocityConversionFactor = Units.rotationsToRadians(1.0) * GEAR_RATIO / 60
+        armMotor.encoder.positionConversionFactor = Units.rotationsToRadians(1.0) * GEAR_RATIO / 60
+
+        armMotor.burnFlash()
+    }
+
+    override fun updateInputs(inputs: IntakeIO.Inputs){
+        inputs.position = Rotation2d(armEncoder.position)
```

Comment:
we should probably be keeping track of arm voltage, because we will probably want to have a threshold that if it crosses, we know we have hit a hard stop and can set absolute encoders to that angle. we will want to be logging that

Reply:
Deferred until after merge.